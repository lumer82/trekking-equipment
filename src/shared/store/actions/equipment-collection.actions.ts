import { Action } from '@ngrx/store';
import { EquipmentCollection } from '../reducer/equipment-collection.reducer';

export const EquipmentCollectionActionTypes = {
  ADD: '[EquipmentCollection] Add'
};

export class EquipmentCollectionAddAction implements Action {
  readonly type: string = EquipmentCollectionActionTypes.ADD;

  constructor(readonly payload: EquipmentCollection) {}
}

export type EquipmentCollectionActions =
  EquipmentCollectionAddAction;
