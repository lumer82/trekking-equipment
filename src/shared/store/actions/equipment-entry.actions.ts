import { AddAction } from '../util/add-action';
import { EquipmentCollection } from '../reducer/equipment-collection.reducer';
import { EquipmentEntry } from '../reducer/equipment-entries.reducer';

export const EquipmentEntryActionTypes = {
  ADD: '[EquipmentEntryAction] Add'
};

export class EquipmentEntryAddAction implements AddAction {
  readonly addAction: boolean = true;
  readonly type: string = EquipmentEntryActionTypes.ADD;

  constructor(public readonly collection: EquipmentCollection,
              public readonly payload: EquipmentEntry) {}
}

export type EquipmentEntryActions =
  EquipmentEntryAddAction;
