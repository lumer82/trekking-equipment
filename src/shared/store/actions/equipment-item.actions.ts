import {AddAction} from '../util/add-action';
import {EquipmentCollection} from '../reducer/equipment-collection.reducer';
import {EquipmentEntry} from '../reducer/equipment-entries.reducer';
import {EquipmentItem} from '../reducer/equipment-item.reducer';

export const EquipmentItemActionTypes = {
  ADD: '[EquipmentItem] Add'
};

export class EquipmentItemAddAction implements AddAction {
  readonly addAction: boolean = true;
  readonly type: string = EquipmentItemActionTypes.ADD;

  constructor(public readonly collection: EquipmentCollection,
              public readonly entry: EquipmentEntry,
              public readonly payload: EquipmentItem) {}
}
