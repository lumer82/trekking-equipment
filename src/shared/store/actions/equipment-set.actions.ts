import { Action } from '@ngrx/store';

export const EquipmentSetActionTypes = {
  CREATE_NEW: '[EquipmentSet] Create new'
};

export class EquipmentSetCreateNewAction implements Action {
  readonly type: string = EquipmentSetActionTypes.CREATE_NEW;
}

export type EquipmentSetActions =
  EquipmentSetCreateNewAction;
