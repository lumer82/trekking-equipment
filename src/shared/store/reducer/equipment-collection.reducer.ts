import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { Action } from '@ngrx/store';
import { EquipmentSetActions, EquipmentSetActionTypes } from '../actions/equipment-set.actions';
import { equipmentEntriesReducer, EquipmentEntriesState } from './equipment-entries.reducer';
import {
  EquipmentCollectionActions, EquipmentCollectionActionTypes,
  EquipmentCollectionAddAction
} from '../actions/equipment-collection.actions';
import { AddAction } from '../util/add-action';

export interface EquipmentCollection {
  id: string;
  name: string;
  entries: EquipmentEntriesState;
}

export type EquipmentCollectionsState = EntityState<EquipmentCollection>;

const adapter: EntityAdapter<EquipmentCollection> = createEntityAdapter<EquipmentCollection>();

const initialState: EquipmentCollectionsState = adapter.getInitialState();

export function equipmentCollectionsReducer(state: EquipmentCollectionsState = initialState,
                                            action: EquipmentSetActions | EquipmentCollectionActions): EquipmentCollectionsState {

  const addAction = (<AddAction>action).addAction;
  Object.values(state.entities)
    .filter(e => !addAction || (<AddAction>action).collection.id === e.id)
    .forEach(e => {
      const entries = equipmentEntriesReducer(e.entries, action);
      if (entries !== e.entries) {
        state = adapter.updateOne({id: e.id, changes: {...e, entries}}, state);
      }
    });

  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return initialState;
    case EquipmentCollectionActionTypes.ADD:
      return adapter.addOne((<EquipmentCollectionAddAction> action).payload, state);
    default:
      return state;
  }
}
