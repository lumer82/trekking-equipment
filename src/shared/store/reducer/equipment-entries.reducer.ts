import {createEntityAdapter, EntityAdapter, EntityState} from '@ngrx/entity';
import {Action} from '@ngrx/store';
import {EquipmentSetActionTypes} from '../actions/equipment-set.actions';
import {equipmentItemsReducer, EquipmentItemsState} from './equipment-item.reducer';
import {AddAction} from '../util/add-action';
import {EquipmentEntryActionTypes, EquipmentEntryAddAction} from '../actions/equipment-entry.actions';

export interface EquipmentEntry {
  id: string;
  name: string;
  items: EquipmentItemsState;
}

export interface EquipmentEntriesState extends EntityState<EquipmentEntry> {
  selectedItemId: string | null;
}

const adapter: EntityAdapter<EquipmentEntry> = createEntityAdapter<EquipmentEntry>();

const initialState: EquipmentEntriesState = adapter.getInitialState({selectedItemId: undefined});

export function equipmentEntriesReducer(state: EquipmentEntriesState = initialState, action: Action): EquipmentEntriesState {
  const addAction = (<AddAction>action).addAction;
  Object.values(state.entities)
    .filter(e => !addAction || (<AddAction>action).entry.id === e.id)
    .forEach(e => {
      const items = equipmentItemsReducer(e.items, action);
      if (items !== e.items) {
        state = adapter.updateOne({id: e.id, changes: {...e, items}}, state);
      }
    });

  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return initialState;
    case EquipmentEntryActionTypes.ADD:
      return adapter.addOne((<EquipmentEntryAddAction>action).payload, state);
    default:
      return state;
  }
}
