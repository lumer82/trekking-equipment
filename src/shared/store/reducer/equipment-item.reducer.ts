import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { Action } from '@ngrx/store';
import { EquipmentSetActionTypes } from '../actions/equipment-set.actions';
import {EquipmentItemActionTypes, EquipmentItemAddAction} from '../actions/equipment-item.actions';

export interface EquipmentItem {
  id: string;
  name: string;
  price: number;
  weight: number;
}

export type EquipmentItemsState = EntityState<EquipmentItem>;

const adapter: EntityAdapter<EquipmentItem> = createEntityAdapter<EquipmentItem>();

const initialState: EquipmentItemsState = adapter.getInitialState();

export function equipmentItemsReducer(state: EquipmentItemsState = initialState, action: Action): EquipmentItemsState {
  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return initialState;
    case EquipmentItemActionTypes.ADD:
      return adapter.addOne((<EquipmentItemAddAction>action).payload, state);
    default:
      return state;
  }
}
