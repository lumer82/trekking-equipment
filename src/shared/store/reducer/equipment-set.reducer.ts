import { equipmentCollectionsReducer, EquipmentCollectionsState } from './equipment-collection.reducer';
import { Action } from '@ngrx/store';
import { EquipmentSetActionTypes } from '../actions/equipment-set.actions';

export interface EquipmentSetState {
  _id?: string;
  name: string;
  collections: EquipmentCollectionsState;
}

const initialState: EquipmentSetState = {
  name: '',
  collections: undefined
}

export function equipmentSetReducer(state: EquipmentSetState = initialState, action: Action): EquipmentSetState {
  const collections = equipmentCollectionsReducer(state.collections, action);
  if (collections !== state.collections) {
    state = {...state, collections};
  }

  switch (action.type) {
    case EquipmentSetActionTypes.CREATE_NEW:
      return {...initialState};
    default:
      return state;
  }

}
