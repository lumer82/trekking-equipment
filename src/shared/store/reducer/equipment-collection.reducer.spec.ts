import {
  EquipmentCollection, equipmentCollectionsReducer,
  EquipmentCollectionsState
} from './equipment-collection.reducer';
import { EquipmentCollectionAddAction } from '../actions/equipment-collection.actions';
import { EquipmentEntryAddAction } from '../actions/equipment-entry.actions';
import { EquipmentEntry } from './equipment-entries.reducer';

describe('EquipmentCollectionReducer', () => {
  describe('undefined action and state', () => {
    it('should return the default state', () => {
      const actual = equipmentCollectionsReducer(undefined, {} as any);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('add collection', () => {
    it('should add the collection to the state', () => {
      const id = Math.random() * Math.pow(10, 10) + '';
      const collection: EquipmentCollection = { id, name: 'Test', entries: null };
      const actual = equipmentCollectionsReducer(undefined, new EquipmentCollectionAddAction(collection));
      expect(actual.entities[id]).toBe(collection);
    });
  });

  describe('add entry', () => {
    it('should add an entry to the collection', () => {
      const entry: EquipmentEntry = { id: '98765', name: 'Entry', items: undefined };
      const collection: EquipmentCollection = { id: '12345', name: 'Test', entries: undefined };
      const state: EquipmentCollectionsState = { ids: ['12345'], entities: { 12345: collection } };
      const actual = equipmentCollectionsReducer(state, new EquipmentEntryAddAction(collection, entry));
      expect(actual.entities['12345'].entries.entities['98765']).toEqual(entry);
      expect(actual).not.toBe(state);
    });

    it('should not add an entry to the wrong collection', () => {
      const entry: EquipmentEntry = { id: '98765', name: 'Entry', items: undefined };
      const collection1: EquipmentCollection = { id: '12345', name: 'Test1', entries: undefined };
      const collection2: EquipmentCollection = { id: '23456', name: 'Test2', entries: undefined };
      const state: EquipmentCollectionsState = { ids: ['12345'], entities: { 12345: collection1, 23456: collection2 } };
      const actual = equipmentCollectionsReducer(state, new EquipmentEntryAddAction(collection1, entry));
      expect(actual.entities[collection2.id].entries).toBeFalsy();
      expect(actual).not.toBe(state);
    });
  });
});
