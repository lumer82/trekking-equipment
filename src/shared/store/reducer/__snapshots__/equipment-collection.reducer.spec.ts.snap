// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EquipmentCollectionReducer undefined action and state should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
}
`;
